powerauth-backend-tests/src/test/java/com/wultra/security/powerauth/test/v31/PowerAuthSignatureTest.java

```diff
@@ -0,0 +1,810 @@
+
+        // Split the offline data into individual lines
+        String[] parts = offlineData.split("\n");
+
+        // Extract last line which contains information about key and ECDSA signature
+        String lastLine = parts[parts.length-1];
+
+        // 1 = KEY_SERVER_PRIVATE was used to sign data (personalized offline signature)
+        assertEquals("1", lastLine.substring(0, 1));
+
+        // The remainder of last line is Base64 encoded ECDSA signature
+        String ecdsaSignature = lastLine.substring(1);
+        final byte[] serverPublicKeyBytes = BaseEncoding.base64().decode((String) model.getResultStatusObject().get("serverPublicKey"));
+        final ECPublicKey serverPublicKey = (ECPublicKey) config.getKeyConversion().convertBytesToPublicKey(serverPublicKeyBytes);
+
+        // Prepare offline data without signature
+        String offlineDataWithoutSignature = offlineData.substring(0, offlineData.length() - ecdsaSignature.length());
+
+        // Validate ECDSA signature of data using server public key
+        assertTrue(signatureUtils.validateECDSASignature(offlineDataWithoutSignature.getBytes(StandardCharsets.UTF_8), BaseEncoding.base64().decode(ecdsaSignature), serverPublicKey));
+
+        // Prepare data for PowerAuth signature
+        String dataForSignature = operationId + "&" + operationData;
+
+        // Prepare normalized data for signature
+        String signatureBaseString = PowerAuthHttpBody.getSignatureBaseString("POST", "/operation/authorize/offline", BaseEncoding.base64().decode(nonce), dataForSignature.getBytes(StandardCharsets.UTF_8));
+
+        // Prepare keys
+        byte[] signaturePossessionKeyBytes = BaseEncoding.base64().decode((String) model.getResultStatusObject().get("signaturePossessionKey"));
+        byte[] signatureKnowledgeKeySalt = BaseEncoding.base64().decode((String) model.getResultStatusObject().get("signatureKnowledgeKeySalt"));
+        byte[] signatureKnowledgeKeyEncryptedBytes = BaseEncoding.base64().decode((String) model.getResultStatusObject().get("signatureKnowledgeKeyEncrypted"));
+
+        // Get the signature keys
+        SecretKey signaturePossessionKey = config.getKeyConversion().convertBytesToSharedSecretKey(signaturePossessionKeyBytes);
+        SecretKey signatureKnowledgeKey = EncryptedStorageUtil.getSignatureKnowledgeKey(config.getPassword().toCharArray(), signatureKnowledgeKeyEncryptedBytes, signatureKnowledgeKeySalt, new KeyGenerator());
+
+        // Put keys into a list
+        List<SecretKey> signatureKeys = new ArrayList<>();
+        signatureKeys.add(signaturePossessionKey);
+        signatureKeys.add(signatureKnowledgeKey);
+
+        // Calculate signature of normalized signature base string with 'offline' as application secret
+        String signature = signatureUtils.computePowerAuthSignature((signatureBaseString + "&offline").getBytes(StandardCharsets.UTF_8), signatureKeys, CounterUtil.getCtrData(model, stepLogger), PowerAuthSignatureFormat.DECIMAL);
```

Comment:
Shouldn't we also add some `BASE64` signature tests?

Reply:
This test function is probably redundant, because the offline signature validation remains the same, since protocol version `2.0`. All other signature validations in `v31` package, are based on BASE64 signatures. So I don't think that we have to explicitly test that the signature is BASE64. Especially on this level of tests.